src/power.jl

```diff
@@ -7,156 +7,251 @@
-    coeffs = Array{eltype(a)}(a.order+1)
-    coeffs[1] = a[1]^2
-    @inbounds for k = 1:a.order
-        coeffs[k+1] = squareHomogCoef(k, a.coeffs)
+"""
+    square(a::AbstractSeries) --> typeof(a)
+
+Return `a^2`; see [TaylorSeries.sqr!](@ref).
+""" square
+
+for T in (:Taylor1, :TaylorN)
+    @eval function square(a::$T)
+        if $T == Taylor1
+            c = Taylor1(Array{eltype(a)}(a.order+1))
+        else
+            c = TaylorN( zeros(HomogeneousPolynomial{eltype(a)}, a.order) )
+        end
+
+        @inbounds for k = 0:a.order
+            sqr!(c, a, k)
+        end
+        return c
     end
-    Taylor1(coeffs,a.order)
 end
 
+function square(a::HomogeneousPolynomial)
+    T = eltype(a)
+
+    order = 2*a.order
+    order > get_order() && return HomogeneousPolynomial([zero(T)], get_order())
+
+    res = HomogeneousPolynomial([zero(T)], order)
+    sqr!(res, a)
+    return res
+end
+
+
 # Homogeneous coefficients for square
 doc"""
-    squareHomogCoef(kcoef, ac)
+    sqr!(c, a, k::Int) --> nothing
 
-Compute the `k-th` expansion coefficient of $c = a^2$, given by
+Update the `k-th` expansion coefficient `c[k+1]` of `c = a^2`, for
+both `c` and `a` either `Taylor1` or `TaylorN`.
 
-\begin{eqnarray*}
-c_k &=& 2 \sum_{j=0}^{(k-1)/2} a_{k-j} a_j, \text{ if $k$ is odd, or }  \\\\
-c_k &=& 2 \sum_{j=0}^{(k-2)/2} a_{k-j} a_j + (a_{k/2})^2, \text{ if $k$ is even, }
-\end{eqnarray*}
+The coefficients are given by
+
+\\begin{eqnarray*}
+c_k & = & 2 \\sum_{j=0}^{(k-1)/2} a_{k-j} a_j,
+    \\text{ if k is odd, or }  \\\\
+c_k & = & 2 \\sum_{j=0}^{(k-2)/2} a_{k-j} a_j + (a_{k/2})^2,
+    \\text{ if k is even. }
+\\end{eqnarray*}
+
+""" sqr!
+
+for T = (:Taylor1, :TaylorN)
+    @eval function sqr!(c::$T, a::$T, k::Int)
+        if k == 0
+            @inbounds c[1] = a[1]^2
+            return nothing
+        end
 
-with $a$ a `Taylor1` polynomial.
+        if $T == Taylor1
+            @inbounds c[k+1] = zero(c[1])
+        else
+            c[k+1] = HomogeneousPolynomial(zero(c[1][1]), k)
+        end
+
+        kodd = k%2
+        kend = div(k - 2 + kodd, 2)
+        for i = 0:kend
+            @inbounds c[k+1] += a[i+1]*a[k-i+1]
+        end
+        c[k+1] = 2 * c[k+1]
+        kodd == 1 && return nothing
+        @inbounds c[k+1] += a[div(k,2)+1]^2
+
+        return nothing
+    end
+end
+
+
+"""
+    sqr!(c, a)
+
+Return `c = a*a` with no allocation; all parameters are `HomogeneousPolynomial`.
 
-Inputs are the `kcoef`-th coefficient and the vector of the expansion coefficients
-`ac` of `a`.
 """
+function sqr!(c::HomogeneousPolynomial, a::HomogeneousPolynomial)
+    iszero(a) && return nothing
+
+    T = eltype(c)
+    @inbounds num_coeffs_a = size_table[a.order+1]
+    @inbounds num_coeffs  = size_table[c.order+1]
+
+    @inbounds posTb = pos_table[c.order+1]
+
+    for na = 1:num_coeffs_a
+        @inbounds ca = a[na]
+        ca == zero(T) && continue
+        @inbounds inda = index_table[a.order+1][na]
```

Comment:
Try given a name to `index_table[a.order+1]` outside the loop. I expect that would be a bit more efficient. (Just like with `posTb`.)